conv_ftl: add procfs on conv_ftl

hyeongtakji opened the pull request:
This patch adds procfs on `conv_ftl` as follows:

```
  $ tree /proc/nvmev/
  /proc/nvmev/
  ├── debug
  ├── ftls
  │   ├── ftl0
  │   │   └── free_line_cnt
  │   ├── ftl1
  │   │   └── free_line_cnt
  │   ├── ftl2
  │   │   └── free_line_cnt
  │   └── ftl3
  │       └── free_line_cnt
  ├── io_units
  ├── read_times
  ├── stat
  └── write_times

  $ cat /proc/nvmev/ftls/ftl*/free_line_cnt
  8190
  8190
  8190
  8190
```

Please note that `free_line_cnt` is just a sample.

hyeongtakji committed:
conv_ftl: add procfs on conv_ftl

This patch is a preparation for letting users to check host and nand
write counts via procfs.

The procfs entries added are:

  $ tree /proc/nvmev
  /proc/nvmev/
  ├── debug
  ├── ftls
  │   ├── ftl0
  │   ├── ftl1
  │   ├── ftl2
  │   └── ftl3
  ├── io_units
  ├── read_times
  ├── stat
  └── write_times

Signed-off-by: Hyeongtak Ji <hyeongtak.ji@gmail.com>

sejun000 commented on conv_ftl.c:
```diff
@@ -2,10 +2,41 @@
 #include "nvmev.h"
 #include "conv_ftl.h"
 
+// proc fs read function to show free_line_cnt
```
고생하셨습니다. 
이왕하시는김에 ... 요것도 ..(아니면 시간나실때 나중에 하셔도 됩니다) 
```
diff --git a/conv_ftl.c b/conv_ftl.c
index 1d0b7ae..9708ac2 100644
--- a/conv_ftl.c
+++ b/conv_ftl.c
@@ -6,6 +6,9 @@
 #include "nvmev.h"
 #include "conv_ftl.h"
 
+uint64_t write_count;
+uint64_t nand_write_count;
+
 static uint32_t get_nr_streams(struct conv_ftl *conv_ftl)
 {
        return conv_ftl->ssd->sp.nstreams;
@@ -1015,6 +1018,7 @@ static bool conv_write(struct nvmev_ns *ns, struct nvmev_request *req, struct nv
                uint64_t local_lpn;
                uint64_t nsecs_completed = 0;
                struct ppa ppa;
+               write_count++;


diff --git a/ssd.c b/ssd.c
index 011f186..b76dce5 100644
--- a/ssd.c
+++ b/ssd.c

+extern uint64_t nand_write_count;
+
 uint64_t ssd_advance_nand(struct ssd *ssd, struct nand_cmd *ncmd)
 {
        int c = ncmd->cmd;
@@ -429,6 +433,7 @@ uint64_t ssd_advance_nand(struct ssd *ssd, struct nand_cmd *ncmd)
                nand_etime = nand_stime + spp->pg_wr_lat;
                lun->next_lun_avail_time = nand_etime;
                completed_time = nand_etime;
+               nand_write_count +=  ncmd->xfer_size / spp->pgsz;
                break;
```

hyeongtakji replied:
패치 감사합니다. 올려주신 것 포함해서 올리겠습니다.

hyeongtakji replied:
procfs로 보기에 유용한 것 올려주셨으니 사실 free_line_cnt는 예시로 올려둔거라 지우겠습니다.

hyeongtakji replied:
패치 쪼개서 host write, nand write 보여주도록 변경했습니다. 커밋 메시지의

> Signed-off-by: Sejun Kwon <sejun000@gmail.com>

는 제가 임의로 추가했는데 원하시면 지우겠습니다.

sejun000 replied:
감사합니다~

SejunKwon committed:
conv_ftl: add host and nand write counts via procfs

WAF is a crucial factor for assessing FTL performance.

This commit modifies conv_ftl to track host and nand write counts,
exposing them through procfs.

  $ tree /proc/nvmev/
  /proc/nvmev/
  ├── debug
  ├── ftls
  │   ├── ftl0
  │   ├── ftl1
  │   ├── ftl2
  │   ├── ftl3
  │   ├── nand_write_count
  │   └── write_count
  ├── io_units
  ├── read_times
  ├── stat
  └── write_times

Signed-off-by: Sejun Kwon <sejun000@gmail.com>
Signed-off-by: Hyeongtak Ji <hyeongtak.ji@gmail.com>

hyeongtakji commented:
제가 돌려봤을 때는 잘 동작하는 것 같은데, 혹시 문제 있으면 알려주세요.

hyeongtakji commented:
커밋 메시지의 사소한 오타가 있어서 수정했습니다.
